src/cephadm/cephadm

```diff
@@ -1200,117 +1235,53 @@ def call(ctx: CephadmContext,
-                    message = message_b
-                else:
-                    assert False
-                if stop and message:
-                    # process has terminated, but have more to read still, so not stopping yet
-                    # (os.read returns '' when it encounters EOF)
-                    stop = False
-                if not message:
-                    continue
-                if fd == process.stdout.fileno():
-                    out += message
-                    message = out_buffer + message
-                    lines = message.split('\n')
-                    out_buffer = lines.pop()
-                    for line in lines:
-                        if verbosity == CallVerbosity.VERBOSE:
-                            logger.info(desc + 'stdout ' + line)
-                        elif verbosity != CallVerbosity.SILENT:
-                            logger.debug(desc + 'stdout ' + line)
-                elif fd == process.stderr.fileno():
-                    err += message
-                    message = err_buffer + message
-                    lines = message.split('\n')
-                    err_buffer = lines.pop()
-                    for line in lines:
-                        if verbosity == CallVerbosity.VERBOSE:
-                            logger.info(desc + 'stderr ' + line)
-                        elif verbosity != CallVerbosity.SILENT:
-                            logger.debug(desc + 'stderr ' + line)
-                else:
-                    assert False
-            except (IOError, OSError):
-                pass
-        if verbosity == CallVerbosity.VERBOSE:
-            logger.debug(desc + 'profile rt=%s, stop=%s, exit=%s, reads=%s'
-                % (time.time()-start_time, stop, process.poll(), reads))
-
-    returncode = process.wait()
-
-    if out_buffer != '':
-        if verbosity == CallVerbosity.VERBOSE:
-            logger.info(desc + 'stdout ' + out_buffer)
-        elif verbosity != CallVerbosity.SILENT:
-            logger.debug(desc + 'stdout ' + out_buffer)
-    if err_buffer != '':
-        if verbosity == CallVerbosity.VERBOSE:
-            logger.info(desc + 'stderr ' + err_buffer)
-        elif verbosity != CallVerbosity.SILENT:
-            logger.debug(desc + 'stderr ' + err_buffer)
 
+    async def run_with_timeout():
+        loop = asyncio.get_event_loop()
+        proc_exited = loop.create_future()
+        protocol_factory = partial(StreamReaderProto,
+                                   proc_exited,
+                                   prefix, verbosity)
+        transport, protocol = await loop.subprocess_exec(
+            protocol_factory,
+            *command,
+            close_fds=True,
+            **kwargs)
+        proc_transport = cast(asyncio.SubprocessTransport, transport)
+        proc_protocol = cast(StreamReaderProto, protocol)
+        returncode = 0
+        try:
+            if timeout:
+                await asyncio.wait_for(proc_exited, timeout)
```

Comment:
according to the docs, this will also raise a timeout, if the process is blocked. This was previously not the case.